### DISCUSSION TITLE
Request timed out after 3500ms

### DISCUSSION THREAD

**Discussion by jorabra** [nextauthjs/next-auth#3186] [Help] [answered]
react: 17.0.2
next: 11.1.2
next-auth: 4.0.0-beta.6

Provider: custom oidc

We're experiencing "openid-client" timeouts during the authentication process using a less than optimal auth server, and thus need to increase the default timeout slightly. 

Stacktrace:
```
[next-auth][error][OAUTH_CALLBACK_ERROR]
https://next-auth.js.org/errors#oauth_callback_error outgoing request timed out after 3500ms {
  error: {
    message: 'outgoing request timed out after 3500ms',
    stack: 'RPError: outgoing request timed out after 3500ms\n' +
      '    at <next_js_app>/node_modules/next-auth/node_modules/openid-client/lib/helpers/request.js:129:13\n' +
      '    at async Client.grant (<next_js_app>/node_modules/next-auth/node_modules/openid-client/lib/client.js:1256:22)\n' +
      '    at async Client.callback (<next_js_app>/node_modules/next-auth/node_modules/openid-client/lib/client.js:450:24)\n' +
      '    at async oAuthCallback (<next_js_app>/node_modules/next-auth/core/lib/oauth/callback.js:103:16)\n' +
      '    at async Object.callback (<next_js_app>/node_modules/next-auth/core/routes/callback.js:50:11)\n' +
      '    at async NextAuthHandler (<next_js_app>/node_modules/next-auth/core/index.js:103:28)\n' +
      '    at async NextAuthNextHandler (<next_js_app>/node_modules/next-auth/next/index.js:40:7)\n' +
      '    at async __WEBPACK_DEFAULT_EXPORT__ (webpack-internal:///./pages/api/auth/[...nextauth].ts:165:86)\n' +
      '    at async Object.apiResolver (<next_js_app>/node_modules/next/dist/server/api-utils.js:101:9)\n' +
      '    at async DevServer.handleApiRequest (<next_js_app>/node_modules/next/dist/server/next-server.js:770:9)',
    name: 'RPError'
  },
  providerId: 'someProviderId',
  message: 'outgoing request timed out after 3500ms'
}
```

It's an error message from the "openid-client" package which has a default timeout value of 3500ms. The solution would be to set an increased default timeout value: https://github.com/panva/node-openid-client/tree/main/docs#customizing

Specifically:
```typescript
import { custom } from 'openid-client';

custom.setHttpOptionsDefaults({
  timeout: 5000,
});
```

I don't seem to be able to trigger a `setDefaults` in 'openid-client' when running the `setHttpOptionsDefaults` in various places of my Next.js application. Not even from a custom server. It's probably because it's executed in the wrong context. That leads me to believe that we need to able to set this through the Next Auth setup, preferably in the `NextAuthOptions`.

I've tested calling `setHttpOptionsDefaults` from [https://github.com/nextauthjs/next-auth/blob/v4.0.0-beta.6/src/core/lib/oauth/client.ts](https://github.com/nextauthjs/next-auth/blob/v4.0.0-beta.6/src/core/lib/oauth/client.ts) with success.

Are there any current way of setting the "openid-client" timeout in Next Auth that I have missed?

If there are no current solution for this I would be happy to make a PR if this sounds reasonable.

**Comment by balazsorban44** [comment]
Right... Similar to what we had to do here: https://github.com/nextauthjs/next-auth/blob/043b252940e262361dbe62816793ccc890ca71f2/src/core/lib/oauth/client.ts#L45

I'll think of a way to make this easier to configure.

**Comment by balazsorban44** [comment, marked as answer]
Opened a PR to expose this option: https://github.com/nextauthjs/next-auth/pull/3188.

We can try to start being restrictive and open up for new options if there is interest later on.

Here is an experimental release if you want to test it out: https://github.com/nextauthjs/next-auth/pull/3188#issuecomment-967042478

**Reply by jorabra** [reply]
Yes, that solves the issue perfectly! Thanks! That was fast 🙂  How long until we see this in the next beta?

**Reply by balazsorban44** [reply]
Hopefully in a few days. v4 stable is also nearing... :shushing_face:

**Reply by AaronConlon** [reply]
Can we set a retry times field? 🤫

**Comment by khuezy** [comment]
@jorabra What did you end up setting the timeout to? I increased it to 10 seconds but it keeps hanging causing my lambda to abort.

**Comment by statusunknown418** [comment]
this is still happening on latest versions

**Reply by devodii** [reply]
Here's what I did.
Go to [next-auth/src/core/lib/oauth/client.ts](https://github.com/nextauthjs/next-auth/blob/043b252940e262361dbe62816793ccc890ca71f2/src/core/lib/oauth/client.ts#L45) inside your node_modules, and set 
 ` client[custom.clock_tolerance]` to a value greater than 10, maybe 100.
 
 You can find it in line 45.

**Reply by Priyanshu1011** [reply]
Thanks for this. It worked!

**Comment by devodii** [comment]
It's almost 2024, (~4yrs after this issue was created), Yet I still experience this error :(

**Comment by arafays** [comment]
![image](https://github.com/nextauthjs/next-auth/assets/6192554/b0ed7a1c-21b1-4341-8257-ea705ff7e1c5)

21st of December 2023

I noticed it is only happening on localhost it is working on live working deployed on netlify 

fixed it by adding timeout 
```ts
GoogleProvider({
  clientId: env.GOOGLE_CLIENT_ID,
  clientSecret: env.GOOGLE_CLIENT_SECRET,
  profile(profile) {
    return {
      id: profile.sub,
      name: `${profile.given_name} ${profile.family_name}`,
      email: profile.email,
      image: profile.picture,
      emailVerified: profile.email_verified,
      role: profile.role ? profile.role : "user",
    };
  },
  httpOptions: {
    timeout: 10000,
  }
}),
```

**Reply by xdrtas** [reply]
This not work for me :(

**Comment by azwdevops** [comment]
I set the timeout under httpOptions to 100000 and it worked for me

**Reply by Jeramy499k** [reply]
sounds crazy to set timeout to 100k, but it works fine haha.

**Reply by elaine202310** [reply]
but still not work for me

**Reply by glovebx** [reply]
still not work for me
